Drop next() as a function parameter type on listeners

### Description

Describe your issue here.

### What type of issue is this? (place an `x` in one of the `[ ]`)
- [x] bug
- [ ] enhancement (feature request)
- [x] question
- [ ] documentation related
- [ ] example code related
- [ ] testing related
- [ ] discussion

### Requirements (place an `x` in each of the `[ ]`)
* [x] I've read and understood the [Contributing guidelines](https://github.com/slackapi/bolt/blob/main/.github/contributing.md) and have done my best effort to follow them.
* [x] I've read and agree to the [Code of Conduct](https://slackhq.github.io/code-of-conduct).
* [x] I've searched for any related issues and avoided creating a duplicate issue.

---

### Bug Report

Filling out the following details about bugs will help us solve your issue sooner.

#### Reproducible in:

package version: 3.11.0

node version: 16.15.0

OS version(s): Docker - node:16-alpine

#### Steps to reproduce:

1. Create an action listener on an App instance and call next()
```typescript
            // Log action_ids of all actions when in Debug mode
            bot.appInstance?.action(/.*/, async ({ action, logger, next }) => {
                if ('action_id' in action) {
                    logger.debug(`Action Hit(${botName}): ${action.action_id}`);
                }
                await next();
            });
```


#### Expected result:

Either TypeScript does not provide next as a valid parameter, or next to exist.

#### Actual result:

Error:
```
[ERROR]  bolt-app UnknownError: next is not a function
    at asCodedError (/workspace/.yarn/cache/@slack-bolt-npm-3.11.0-1183767d75-25491d656e.zip/node_modules/@slack/bolt/src/errors.ts:56:10)
    at App.handleError (/workspace/.yarn/cache/@slack-bolt-npm-3.11.0-1183767d75-25491d656e.zip/node_modules/@slack/bolt/src/App.ts:1049:37)
    at App.processEvent (/workspace/.yarn/cache/@slack-bolt-npm-3.11.0-1183767d75-25491d656e.zip/node_modules/@slack/bolt/src/App.ts:1032:19)
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at async SocketModeClient.<anonymous> (/workspace/.yarn/cache/@slack-bolt-npm-3.11.0-1183767d75-25491d656e.zip/node_modules/@slack/bolt/src/receivers/SocketModeReceiver.ts:218:9) {
  code: 'slack_bolt_unknown_error',
  original: TypeError: next is not a function
      at /workspace/src/core/botManager.ts:64:23
      at /workspace/.yarn/cache/@slack-bolt-npm-3.11.0-1183767d75-25491d656e.zip/node_modules/@slack/bolt/src/App.ts:1006:27
      at invokeMiddleware (/workspace/.yarn/cache/@slack-bolt-npm-3.11.0-1183767d75-25491d656e.zip/node_modules/@slack/bolt/src/middleware/process.ts:32:12)
      at next (/workspace/.yarn/cache/@slack-bolt-npm-3.11.0-1183767d75-25491d656e.zip/node_modules/@slack/bolt/src/middleware/process.ts:24:21)
      at Array.<anonymous> (/workspace/.yarn/cache/@slack-bolt-npm-3.11.0-1183767d75-25491d656e.zip/node_modules/@slack/bolt/src/middleware/builtin.ts:197:11)
      at invokeMiddleware (/workspace/.yarn/cache/@slack-bolt-npm-3.11.0-1183767d75-25491d656e.zip/node_modules/@slack/bolt/src/middleware/process.ts:23:47)
      at next (/workspace/.yarn/cache/@slack-bolt-npm-3.11.0-1183767d75-25491d656e.zip/node_modules/@slack/bolt/src/middleware/process.ts:24:21)
      at Array.onlyActions (/workspace/.yarn/cache/@slack-bolt-npm-3.11.0-1183767d75-25491d656e.zip/node_modules/@slack/bolt/src/middleware/builtin.ts:42:9)
      at invokeMiddleware (/workspace/.yarn/cache/@slack-bolt-npm-3.11.0-1183767d75-25491d656e.zip/node_modules/@slack/bolt/src/middleware/process.ts:23:47)
      at processMiddleware (/workspace/.yarn/cache/@slack-bolt-npm-3.11.0-1183767d75-25491d656e.zip/node_modules/@slack/bolt/src/middleware/process.ts:35:10)
```

#### Attachments:

Logs, screenshots, screencast, sample project, funny gif, etc.

--- a/src/App-built-in-middleware.spec.ts
+++ b/src/App-built-in-middleware.spec.ts
@@ -366,6 +366,19 @@ describe('App built-in middleware and mechanism', () => {
       assert.throws(() => app.event('message.channels', async () => {}), 'Although the document mentions');
       assert.throws(() => app.event(/message\..+/, async () => {}), 'Although the document mentions');
     });
+
+    // https://github.com/slackapi/bolt-js/issues/1457
+    it('should not cause a runtime exception if the last listener middleware invokes next()', async () => new Promise((resolve, reject) => {
+      app.event('app_mention', async ({ next }) => {
+        try {
+          await next();
+          resolve();
+        } catch (e) {
+          reject(e);
+        }
+      });
+      fakeReceiver.sendEvent(createDummyReceiverEvent('app_mention'));
+    }));
   });
 
   describe('middleware and listener arguments', () => {

--- a/src/App.ts
+++ b/src/App.ts
@@ -1143,7 +1143,7 @@ export default class App<AppCustomContext extends StringIndexed = StringIndexed>
             // Copy the array so modifications don't affect the original
             const listenerMiddleware = [...origListenerMiddleware];
 
-            // Don't process the last item in the listenerMiddleware array - it shouldn't get a next fn
+            // Don't process the last item in the listenerMiddleware array - it will be passed a no-op next fn
             const listener = listenerMiddleware.pop();
 
             if (listener === undefined) {
@@ -1156,13 +1156,13 @@ export default class App<AppCustomContext extends StringIndexed = StringIndexed>
               client,
               this.logger,
               // When all the listener middleware are done processing,
-              // `listener` here will be called without a `next` execution
+              // `listener` here will be called with a noop `next` fn
               async () => listener({
                 ...(listenerArgs as AnyMiddlewareArgs),
                 context,
                 client,
                 logger: this.logger,
-                // `next` is already set in the outer processMiddleware
+                next: () => {},
               } as AnyMiddlewareArgs & AllMiddlewareArgs),
             );
           });

